File: windows/Marquee.Core/Models/TitleModels.cs
using System.Globalization;

namespace Marquee.Core.Models;

// The title page: detail, seasons and episodes (api-v1.md section 3), and the
// response to requesting one (the request-creation part of section 7). The
// library and action-state blocks it embeds are in LibraryModels.cs. Mirrors
// mac/Marquee/API/Models/TitleModels.swift; the nested Swift types are
// top-level records here because C# won't let a property and a nested type
// share a name (`facts: Facts`).

/// <summary>
/// The sidebar and the line under the title. Date labels are pre-formatted
/// in the server's locale and time zone.
/// </summary>
public sealed record TitleFacts
{
    public int? RuntimeMinutes { get; init; }

    /// <summary>"2h 16m" for movies, "~42m/episode" (average) for TV; both null when TMDb has no runtime.</summary>
    public string? RuntimeLabel { get; init; }

    /// <summary>TMDb vote average × 10.</summary>
    public int? RatingPercent { get; init; }

    /// <summary>At most 3.</summary>
    public required IReadOnlyList<string> Genres { get; init; }

    /// <summary>"2011–2019" for an ended show.</summary>
    public string? YearRange { get; init; }

    /// <summary>TMDb's status, "Returning Series" relabeled "Continuing".</summary>
    public string? StatusLabel { get; init; }

    /// <summary>TV only.</summary>
    public string? Network { get; init; }

    /// <summary>"Release Date" (movie) / "First Air Date" (TV).</summary>
    public string? ReleaseDateLabel { get; init; }

    public DateOnly? NextAirDate { get; init; }
    public string? NextAirDateLabel { get; init; }
    public string? OriginalLanguage { get; init; }
    public string? OriginalLanguageLabel { get; init; }
    public ProductionCountry? ProductionCountry { get; init; }

    /// <summary>"Currently Streaming On": US flat-rate providers.</summary>
    public required IReadOnlyList<WatchProvider> WatchProviders { get; init; }
}

public sealed record ProductionCountry
{
    /// <summary>ISO 3166-1, e.g. <c>"US"</c>.</summary>
    public required string Code { get; init; }

    public required string Name { get; init; }

    /// <summary>🇺🇸</summary>
    public required string Flag { get; init; }
}

public sealed record WatchProvider
{
    public required string Name { get; init; }
    public ImageRef? LogoPath { get; init; }
}

/// <summary>Director + Screenplay/Writer (movies) or Creator + Executive Producer (TV), max 6.</summary>
public sealed record TitleCredit
{
    public required string Role { get; init; }
    public required string Name { get; init; }
}

public sealed record TitleLinks
{
    public string? TrailerYoutubeKey { get; init; }
    public string? ImdbId { get; init; }
    public int? TvdbId { get; init; }
    public string? FacebookId { get; init; }
    public string? InstagramId { get; init; }
    public string? TwitterId { get; init; }

    /// <summary>The ordered button row after "▶ Trailer".</summary>
    public required IReadOnlyList<ExternalLink> External { get; init; }

    /// <summary><c>https://www.youtube.com/watch?v={trailerYoutubeKey}</c>.</summary>
    public Uri? TrailerUrl =>
        TrailerYoutubeKey.NonBlank() is { } key
        && Uri.TryCreate("https://www.youtube.com/watch?v=" + Uri.EscapeDataString(key), UriKind.Absolute, out var url)
            ? url
            : null;
}

public sealed record ExternalLink
{
    /// <summary>"IMDb", "TheTVDB", "Instagram", "X / Twitter", "Facebook".</summary>
    public required string Label { get; init; }

    public required string Url { get; init; }

    public Uri? Link => Uri.TryCreate(Url, UriKind.Absolute, out var link) ? link : null;
}

/// <summary>A season row in the "Episodes" accordion.</summary>
public sealed record SeasonSummary
{
    public required int SeasonNumber { get; init; }
    public required string Name { get; init; }
    public required int EpisodeCount { get; init; }
    public DateOnly? AirDate { get; init; }
    public ImageRef? PosterPath { get; init; }

    /// <summary>Sonarr episode-file counts; null when Sonarr doesn't track the show.</summary>
    public int? Have { get; init; }

    public int? Total { get; init; }

    /// <summary>
    /// Season requests (null from an older server): whether Sonarr monitors
    /// this season; null when Sonarr doesn't track the show or isn't connected.
    /// </summary>
    public bool? Monitored { get; init; }

    /// <summary>In one of the viewer's pending or approved requests for this title.</summary>
    public bool? Requested { get; init; }

    /// <summary>The season picker offers a checkbox: not complete, not monitored, not already requested by this viewer.</summary>
    public bool? Requestable { get; init; }

    public int Id => SeasonNumber;

    /// <summary>The season picker's row: a checkbox, or the reason there isn't one.</summary>
    public SeasonRequestState RequestState =>
        Requestable == true ? SeasonRequestState.Requestable
        : IsComplete ? SeasonRequestState.InLibrary
        : Monitored == true ? SeasonRequestState.Monitored
        : Requested == true ? SeasonRequestState.Requested
        : SeasonRequestState.Unavailable;

    /// <summary>The "3/10" badge, null when Sonarr doesn't track the show.</summary>
    public string? CompletenessLabel =>
        Have is { } have && Total is { } total
            ? $"{have.ToString(CultureInfo.InvariantCulture)}/{total.ToString(CultureInfo.InvariantCulture)}"
            : null;

    /// <summary>Green badge: every episode has a file.</summary>
    public bool IsComplete => Have is { } have && Total is { } total && total > 0 && have >= total;
}

/// <summary>A row of the season picker.</summary>
public enum SeasonRequestState
{
    /// <summary>A checkbox.</summary>
    Requestable,

    /// <summary>"In library": every episode has a file.</summary>
    InLibrary,

    /// <summary>"Monitored": Sonarr is already after it.</summary>
    Monitored,

    /// <summary>"Requested": in one of your requests.</summary>
    Requested,

    /// <summary>Not offered, for no reason the server spelled out.</summary>
    Unavailable,
}

public static class SeasonRequestStateExtensions
{
    /// <summary>The tag in place of the checkbox; empty for a checkbox row or an unexplained one.</summary>
    public static string Tag(this SeasonRequestState state) => state switch
    {
        SeasonRequestState.InLibrary => "In library",
        SeasonRequestState.Monitored => "Monitored",
        SeasonRequestState.Requested => "Requested",
        _ => "",
    };
}

/// <summary>What the title page's Request button does.</summary>
public enum TitleRequestAction
{
    /// <summary>No Request button.</summary>
    None,

    /// <summary><c>POST …/request</c> with no body: a movie, a whole series, or a server too old to take seasons.</summary>
    WholeSeries,

    /// <summary>"Request" opens the season picker.</summary>
    PickSeasons,

    /// <summary>"Request more seasons" (the show is already tracked or partly requested) opens the season picker.</summary>
    PickMoreSeasons,
}

public sealed record CastMember
{
    public required int TmdbId { get; init; }
    public required string Name { get; init; }
    public string? Character { get; init; }
    public ImageRef? ProfilePath { get; init; }

    /// <summary>Billing order.</summary>
    public required int Order { get; init; }

    public required bool Favorited { get; init; }

    public int Id => TmdbId;
}

/// <summary>A TMDb collection (movies) or a curated TV crossover group.</summary>
public sealed record TitleFranchise
{
    public required string Title { get; init; }

    /// <summary>Null for TV groups; the heading's star favorites <c>/favorites/collection/{collectionId}</c>.</summary>
    public int? CollectionId { get; init; }

    public bool? CollectionFavorited { get; init; }

    /// <summary>Oldest first, with status, favorited, requested, canQuickAdd, canRequest.</summary>
    public required IReadOnlyList<TitleCard> Items { get; init; }

    /// <summary>The admin's "Add all N missing" set (empty for members); add each with <c>POST /titles/{type}/{id}/add</c>.</summary>
    public required IReadOnlyList<TitleId> AddAllMissing { get; init; }
}

/// <summary><c>GET /titles/{type}/{tmdbId}</c>: everything the title page renders.</summary>
public sealed record TitleDetail
{
    public required MediaType MediaType { get; init; }
    public required int TmdbId { get; init; }
    public int? TvdbId { get; init; }
    public string? ImdbId { get; init; }
    public required string Name { get; init; }
    public string? Overview { get; init; }
    public string? Tagline { get; init; }
    public ImageRef? PosterPath { get; init; }
    public ImageRef? BackdropPath { get; init; }
    public string? Year { get; init; }

    /// <summary>Raw TMDb release / first-air date.</summary>
    public DateOnly? ReleaseDate { get; init; }

    /// <summary>TMDb's own status, e.g. "Released", "Returning Series".</summary>
    public string? TmdbStatus { get; init; }

    public required TitleFacts Facts { get; init; }
    public required IReadOnlyList<TitleCredit> Credits { get; init; }
    public required IReadOnlyList<string> Keywords { get; init; }
    public required TitleLinks Links { get; init; }
    public required TitleLibraryInfo Library { get; init; }

    /// <summary>Decides the action area under the title (see <see cref="TitleViewerState"/>).</summary>
    public required TitleViewerState Viewer { get; init; }

    /// <summary>TV only: seasons with episodes, newest first. Load episodes with <c>Titles.SeasonAsync</c>.</summary>
    public required IReadOnlyList<SeasonSummary> Seasons { get; init; }

    /// <summary>Top 20 by billing order.</summary>
    public required IReadOnlyList<CastMember> Cast { get; init; }

    public TitleFranchise? Franchise { get; init; }

    /// <summary>Heading "Studio".</summary>
    public required IReadOnlyList<CompanyCard> Studios { get; init; }

    /// <summary>TMDb recommendations, heading "More like this".</summary>
    public required IReadOnlyList<TitleCard> Similar { get; init; }

    public TitleId Id => new(MediaType, TmdbId);

    /// <summary>
    /// What the action area's Request button does for this viewer. A server
    /// from before season requests doesn't send <c>canRequestSeasons</c>, so
    /// it keeps today's whole-series Request.
    /// </summary>
    public TitleRequestAction RequestAction
    {
        get
        {
            if (Viewer.AlreadyRequested)
            {
                return TitleRequestAction.None;
            }
            var seasonsOffered = MediaType == MediaType.Tv
                && Viewer.CanRequestSeasons == true
                && Seasons.Any(season => season.RequestState == SeasonRequestState.Requestable);
            if (Viewer is { CanRequest: true, RequestStatus: null })
            {
                return seasonsOffered ? TitleRequestAction.PickSeasons : TitleRequestAction.WholeSeries;
            }
            // "More" only when some of the show is already in the library or on
            // its way, as on the website; an approved request for a show that
            // isn't there yet still just reads "Request".
            if (!seasonsOffered)
            {
                return TitleRequestAction.None;
            }
            return Library.Status == LibraryStatus.Untracked ? TitleRequestAction.PickSeasons : TitleRequestAction.PickMoreSeasons;
        }
    }

    /// <summary>
    /// The same title with a fresh <c>library</c> + <c>viewer</c> from
    /// <c>Titles.StatusAsync</c>, for updating the page after add / request /
    /// monitor / favorite without re-fetching the whole thing.
    /// </summary>
    public TitleDetail Updating(TitleStatus status) => this with { Library = status.Library, Viewer = status.Viewer };
}

/// <summary><c>GET /titles/tv/{tmdbId}/seasons/{n}</c>: one expanded accordion row.</summary>
public sealed record SeasonEpisodes
{
    public required int TmdbId { get; init; }
    public required int SeasonNumber { get; init; }

    /// <summary>Empty: "No episode data for this season."</summary>
    public required IReadOnlyList<Episode> Episodes { get; init; }
}

public sealed record Episode
{
    /// <summary>The website truncates episode overviews at this many characters.</summary>
    public const int OverviewLimit = 220;

    /// <summary>TMDb episode id.</summary>
    public required int Id { get; init; }

    public required int EpisodeNumber { get; init; }
    public required string Name { get; init; }
    public string? Overview { get; init; }
    public DateOnly? AirDate { get; init; }
    public ImageRef? StillPath { get; init; }

    /// <summary>Sonarr's "Have it" / "Missing"; null when Sonarr doesn't track the show.</summary>
    public bool? HasFile { get; init; }

    /// <summary>"S01E03".</summary>
    public string Code(int season) =>
        $"S{season.ToString("00", CultureInfo.InvariantCulture)}E{EpisodeNumber.ToString("00", CultureInfo.InvariantCulture)}";

    /// <summary>The overview cut the way the website's line clamp does, with a trailing "…".</summary>
    public string? ShortOverview => Overview == null ? null : Truncate(Overview, OverviewLimit);

    /// <summary>Cut at <paramref name="limit"/> characters with a trailing "…" (the website's line clamps).</summary>
    internal static string Truncate(string text, int limit) =>
        text.Length <= limit ? text : text[..limit].Trim() + "…";
}

/// <summary>
/// <c>POST /titles/{type}/{tmdbId}/request</c> response (the Mac's
/// <c>RequestCreated</c>; named for the title page here so it can't collide
/// with the requests area's own models).
/// </summary>
public sealed record TitleRequestCreated
{
    public required bool Ok { get; init; }
    public required Guid RequestId { get; init; }
}

/// <summary><c>POST /titles/tv/{tmdbId}/request</c> body for some seasons; the whole series sends no body.</summary>
public sealed record SeasonRequestBody(IReadOnlyList<int> Seasons);

/// <summary>
/// The season picker's selection: only requestable seasons can be picked,
/// and "Select all" toggles all of them. The Mac's <c>SeasonPickerSelection</c>.
/// </summary>
public sealed class SeasonPickerSelection
{
    private readonly HashSet<int> selected = [];

    public SeasonPickerSelection(IEnumerable<SeasonSummary> seasons)
    {
        Requestable = seasons
            .Where(season => season.RequestState == SeasonRequestState.Requestable)
            .Select(season => season.SeasonNumber)
            .ToList();
    }

    /// <summary>The seasons with a checkbox, in the picker's order.</summary>
    public IReadOnlyList<int> Requestable { get; }

    public bool AllSelected => Requestable.Count > 0 && Requestable.All(selected.Contains);

    public bool IsSelected(int season) => selected.Contains(season);

    public void Set(int season, bool on)
    {
        if (!Requestable.Contains(season))
        {
            return;
        }
        if (on)
        {
            selected.Add(season);
        }
        else
        {
            selected.Remove(season);
        }
    }

    /// <summary>"Select all": everything when anything is unticked, else nothing.</summary>
    public void ToggleAll()
    {
        var all = AllSelected;
        selected.Clear();
        if (!all)
        {
            selected.UnionWith(Requestable);
        }
    }

    /// <summary>What the request sends, sorted.</summary>
    public IReadOnlyList<int> Seasons => selected.Order().ToList();

    /// <summary>"Request 1 season" / "Request 3 seasons"; "Request seasons" while nothing is picked (the button is disabled then), as on the website.</summary>
    public string SubmitTitle => selected.Count == 0
        ? "Request seasons"
        : $"Request {selected.Count.ToString(CultureInfo.CurrentCulture)} {(selected.Count == 1 ? "season" : "seasons")}";
}

File: mac/Marquee/API/Models/TitleModels.swift
import Foundation

// The title page: detail, seasons and episodes (api-v1.md §3). The library
// and action-state blocks it embeds are in LibraryModels.swift.

extension API {
    /// `GET /titles/{type}/{tmdbId}`: everything the title page renders.
    struct TitleDetail: Codable, Hashable, Sendable, Identifiable {
        /// The sidebar and the line under the title. Date labels are
        /// pre-formatted in the server's locale and time zone.
        struct Facts: Codable, Hashable, Sendable {
            let runtimeMinutes: Int?
            /// "2h 16m" for movies, "~42m/episode" (average) for TV.
            let runtimeLabel: String?
            /// TMDb vote average × 10.
            let ratingPercent: Int?
            /// At most 3.
            let genres: [String]
            /// "2011–2019" for an ended show.
            let yearRange: String?
            /// TMDb's status, "Returning Series" relabeled "Continuing".
            let statusLabel: String?
            /// TV only.
            let network: String?
            /// "Release Date" (movie) / "First Air Date" (TV).
            let releaseDateLabel: String?
            let nextAirDate: CalendarDay?
            let nextAirDateLabel: String?
            let originalLanguage: String?
            let originalLanguageLabel: String?
            let productionCountry: ProductionCountry?
            /// "Currently Streaming On": US flat-rate providers.
            let watchProviders: [WatchProvider]
        }

        struct ProductionCountry: Codable, Hashable, Sendable {
            /// ISO 3166-1, e.g. `"US"`.
            let code: String
            let name: String
            /// 🇺🇸
            let flag: String
        }

        struct WatchProvider: Codable, Hashable, Sendable {
            let name: String
            let logoPath: ImageRef?
        }

        /// Director + Screenplay/Writer (movies) or Creator + Executive Producer (TV), max 6.
        struct Credit: Codable, Hashable, Sendable {
            let role: String
            let name: String
        }

        struct Links: Codable, Hashable, Sendable {
            let trailerYoutubeKey: String?
            let imdbId: String?
            let tvdbId: Int?
            let facebookId: String?
            let instagramId: String?
            let twitterId: String?
            /// The ordered button row after "▶ Trailer".
            let external: [ExternalLink]

            /// `https://www.youtube.com/watch?v={trailerYoutubeKey}`.
            var trailerURL: URL? {
                guard let key = trailerYoutubeKey.nonBlank,
                      let encoded = key.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) else { return nil }
                return URL(string: "https://www.youtube.com/watch?v=\(encoded)")
            }
        }

        struct ExternalLink: Codable, Hashable, Sendable {
            /// "IMDb", "TheTVDB", "Instagram", "X / Twitter", "Facebook".
            let label: String
            let url: String

            var link: URL? { URL(string: url) }
        }

        /// A season row in the "Episodes" accordion.
        struct SeasonSummary: Codable, Hashable, Sendable, Identifiable {
            let seasonNumber: Int
            let name: String
            let episodeCount: Int
            let airDate: CalendarDay?
            let posterPath: ImageRef?
            /// Sonarr episode-file counts; nil when Sonarr doesn't track the show.
            let have: Int?
            let total: Int?
            /// Season requests (nil from an older server). Whether Sonarr
            /// monitors this season; nil when Sonarr doesn't track the show or
            /// isn't connected.
            let monitored: Bool?
            /// In one of the viewer's pending or approved requests for this title.
            let requested: Bool?
            /// The season picker offers a checkbox: not complete, not
            /// monitored, not already requested by this viewer.
            let requestable: Bool?

            var id: Int { seasonNumber }

            /// The season picker's row: a checkbox, or why there isn't one.
            var requestState: SeasonRequestState {
                if requestable == true { return .requestable }
                if isComplete { return .inLibrary }
                if monitored == true { return .monitored }
                if requested == true { return .requested }
                return .unavailable
            }

            /// "1 episode" / "10 episodes".
            var episodeCountLabel: String {
                "\(episodeCount) episode\(episodeCount == 1 ? "" : "s")"
            }

            /// The "3/10" badge, nil when Sonarr doesn't track the show.
            var completenessLabel: String? {
                guard let have, let total else { return nil }
                return "\(have)/\(total)"
            }

            /// Green badge: every episode has a file.
            var isComplete: Bool {
                guard let have, let total else { return false }
                return total > 0 && have >= total
            }
        }

        struct CastMember: Codable, Hashable, Sendable, Identifiable {
            let tmdbId: Int
            let name: String
            let character: String?
            let profilePath: ImageRef?
            /// Billing order.
            let order: Int
            let favorited: Bool

            var id: Int { tmdbId }
        }

        /// A TMDb collection (movies) or a curated TV crossover group.
        struct Franchise: Codable, Hashable, Sendable {
            let title: String
            /// nil for TV groups; the heading's star favorites `/favorites/collection/{collectionId}`.
            let collectionId: Int?
            let collectionFavorited: Bool?
            /// Oldest first, with status, favorited, requested, canQuickAdd, canRequest.
            let items: [TitleCard]
            /// The admin's "Add all N missing" set (empty for members).
            let addAllMissing: [TitleID]
        }

        let mediaType: MediaType
        let tmdbId: Int
        let tvdbId: Int?
        let imdbId: String?
        let name: String
        let overview: String?
        let tagline: String?
        let posterPath: ImageRef?
        let backdropPath: ImageRef?
        let year: String?
        /// Raw TMDb release / first-air date.
        let releaseDate: CalendarDay?
        /// TMDb's own status, e.g. "Released", "Returning Series".
        let tmdbStatus: String?
        let facts: Facts
        let credits: [Credit]
        let keywords: [String]
        let links: Links
        let library: TitleLibraryInfo
        /// Decides the action area under the title (see `TitleViewerState`).
        let viewer: TitleViewerState
        /// TV only: seasons with episodes, newest first. Load episodes with `titles.season`.
        let seasons: [SeasonSummary]
        /// Top 20 by billing order.
        let cast: [CastMember]
        let franchise: Franchise?
        /// Heading "Studio".
        let studios: [CompanyCard]
        /// TMDb recommendations, heading "More like this".
        let similar: [TitleCard]

        var id: TitleID { TitleID(mediaType, tmdbId) }

        /// What the action area's Request button does for this viewer, or nil
        /// when there's no Request button (see `TitleRequestAction`).
        var requestAction: TitleRequestAction? {
            guard !viewer.alreadyRequested else { return nil }
            // A server from before season requests doesn't send
            // `canRequestSeasons`: today's whole-series Request, unchanged.
            let seasonsOffered = mediaType == .tv
                && viewer.canRequestSeasons == true
                && seasons.contains { $0.requestState == .requestable }
            if viewer.canRequest, viewer.requestStatus == nil {
                return seasonsOffered ? .pickSeasons(more: false) : .wholeSeries
            }
            if seasonsOffered {
                // Already tracked, or an earlier request was approved: more
                // seasons of a show that's partly in the library or on its way.
                return .pickSeasons(more: true)
            }
            return nil
        }

        /// The same title with a fresh `library` + `viewer` from `titles.status`,
        /// for updating the page after add / request / monitor / favorite.
        func updating(_ status: TitleStatus) -> TitleDetail {
            TitleDetail(
                mediaType: mediaType, tmdbId: tmdbId, tvdbId: tvdbId, imdbId: imdbId, name: name,
                overview: overview, tagline: tagline, posterPath: posterPath, backdropPath: backdropPath,
                year: year, releaseDate: releaseDate, tmdbStatus: tmdbStatus, facts: facts, credits: credits,
                keywords: keywords, links: links, library: status.library, viewer: status.viewer,
                seasons: seasons, cast: cast, franchise: franchise, studios: studios, similar: similar
            )
        }
    }

    /// A row of the season picker.
    enum SeasonRequestState: Hashable, Sendable {
        /// A checkbox.
        case requestable
        /// "In library": every episode has a file.
        case inLibrary
        /// "Monitored": Sonarr is already after it.
        case monitored
        /// "Requested": in one of your requests.
        case requested
        /// Not offered, for no reason the server spelled out.
        case unavailable

        /// The tag in place of the checkbox.
        var tag: String? {
            switch self {
            case .requestable, .unavailable: nil
            case .inLibrary: "In library"
            case .monitored: "Monitored"
            case .requested: "Requested"
            }
        }
    }

    /// The title page's Request button.
    enum TitleRequestAction: Hashable, Sendable {
        /// `POST …/request` with no body: a movie, a whole series, or a
        /// server too old to take seasons.
        case wholeSeries
        /// Opens the season picker. `more`: the show is already tracked or
        /// partly requested, so the button reads "Request more seasons".
        case pickSeasons(more: Bool)

        var buttonTitle: String {
            switch self {
            case .wholeSeries, .pickSeasons(more: false): "Request"
            case .pickSeasons(more: true): "Request more seasons"
            }
        }
    }

    /// `GET /titles/tv/{tmdbId}/seasons/{n}`: one expanded accordion row.
    struct SeasonEpisodes: Codable, Hashable, Sendable {
        let tmdbId: Int
        let seasonNumber: Int
        /// Empty → "No episode data for this season."
        let episodes: [Episode]
    }

    struct Episode: Codable, Hashable, Sendable, Identifiable {
        /// TMDb episode id.
        let id: Int
        let episodeNumber: Int
        let name: String
        let overview: String?
        let airDate: CalendarDay?
        let stillPath: ImageRef?
        /// Sonarr's "Have it" / "Missing"; nil when Sonarr doesn't track the show.
        let hasFile: Bool?

        /// "S01E03".
        func code(season: Int) -> String {
            "S\(Format.twoDigits(season))E\(Format.twoDigits(episodeNumber))"
        }

        /// The website truncates episode overviews at 220 characters.
        var shortOverview: String? {
            overview.map { $0.truncated(to: 220) }
        }
    }
}

extension String {
    /// Cut at `limit` characters with a trailing "…" (the website's line clamps).
    func truncated(to limit: Int) -> String {
        guard count > limit else { return self }
        return String(prefix(limit)).trimmingCharacters(in: .whitespaces) + "…"
    }
}
